// angular-calendar-scheduler/esm5/modules/scheduler/calendar-scheduler-event-content.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ElementRef, AfterViewInit } from '@angular/core';
var CalendarSchedulerEventContentComponent = /** @class */ (function () {
    function CalendarSchedulerEventContentComponent(hostElement) {
        this.hostElement = hostElement;
    }
    CalendarSchedulerEventContentComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () { _this.maxHeight = _this.eventContainer.clientHeight - 70; }, 0);
    };
    CalendarSchedulerEventContentComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], CalendarSchedulerEventContentComponent.prototype, "event", void 0);
    tslib_1.__decorate([
        Input()
    ], CalendarSchedulerEventContentComponent.prototype, "eventContainer", void 0);
    CalendarSchedulerEventContentComponent = tslib_1.__decorate([
        Component({
            selector: 'calendar-scheduler-event-content',
            template: "\n        <div *ngIf=\"event.content\"\n            class=\"cal-scheduler-event-content\"\n            [style.max-height.px]=\"maxHeight\"\n            [innerHTML]=\"maxHeight && maxHeight >= 30 ? event.content : null\">\n        </div>\n    ",
            host: {
                'class': 'cal-scheduler-event-content-container'
            }
        })
    ], CalendarSchedulerEventContentComponent);
    return CalendarSchedulerEventContentComponent;
}());
export { CalendarSchedulerEventContentComponent };
