// angular-calendar-scheduler/esm5/modules/scheduler/calendar-scheduler-event-actions.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var CalendarSchedulerEventActionsComponent = /** @class */ (function () {
    function CalendarSchedulerEventActionsComponent() {
        this.actions = [];
    }
    CalendarSchedulerEventActionsComponent.prototype.ngOnInit = function () {
        this.setupActions();
    };
    CalendarSchedulerEventActionsComponent.prototype.ngOnChanges = function (changes) {
        if (changes.event) {
            this.setupActions();
        }
    };
    CalendarSchedulerEventActionsComponent.prototype.setupActions = function () {
        this.actions = this.event.isCancelled
            ? this.event.actions.filter(function (a) { return !a.when || a.when === 'cancelled'; })
            : this.event.isDisabled
                ? this.event.actions.filter(function (a) { return !a.when || a.when === 'disabled'; })
                : this.event.actions.filter(function (a) { return !a.when || a.when === 'enabled'; });
    };
    /**
     * @hidden
     */
    CalendarSchedulerEventActionsComponent.prototype.onActionClick = function (mouseEvent, action, event) {
        if (mouseEvent.stopPropagation) {
            mouseEvent.stopPropagation();
        }
        action.onClick(event);
    };
    tslib_1.__decorate([
        Input()
    ], CalendarSchedulerEventActionsComponent.prototype, "event", void 0);
    CalendarSchedulerEventActionsComponent = tslib_1.__decorate([
        Component({
            selector: 'calendar-scheduler-event-actions',
            template: "\n        <span *ngIf=\"event.actions\"\n            class=\"cal-scheduler-event-actions\">\n            <a\n                class=\"cal-scheduler-event-action\"\n                href=\"javascript:;\"\n                *ngFor=\"let action of actions\"\n                (mwlClick)=\"onActionClick($event, action, event)\"\n                [ngClass]=\"action.cssClass\"\n                [innerHtml]=\"action.label\"\n                [title]=\"action.title\">\n            </a>\n        </span>\n    ",
            host: {
                'class': 'cal-scheduler-event-actions-container'
            }
        })
    ], CalendarSchedulerEventActionsComponent);
    return CalendarSchedulerEventActionsComponent;
}());
export { CalendarSchedulerEventActionsComponent };
