// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-hour-segment.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
// WORKAROUND: https://github.com/dherges/ng-packagr/issues/217#issuecomment-339460255
import * as momentImported from 'moment';
const moment = momentImported;
let CalendarSchedulerHourSegmentComponent = class CalendarSchedulerHourSegmentComponent {
    constructor() {
        this.hourSegmentHeight = 58;
        this.showHour = false;
        this.segmentClicked = new EventEmitter();
    }
    ngOnInit() {
        this.title = this.title || moment(this.segment.date).format('dddd L, LT');
    }
    /**
     * @hidden
     */
    onMouseEnter(mouseEvent, segment, event) {
        if (!event.isDisabled && !segment.isDisabled) {
            // Maybe do something
        }
    }
    /**
     * @hidden
     */
    onMouseLeave(mouseEvent, segment, event) {
        if (!event.isDisabled && !segment.isDisabled) {
            // Maybe do something
        }
    }
    /**
     * @hidden
     */
    onSegmentClick(mouseEvent, segment) {
        if (mouseEvent.stopPropagation) {
            mouseEvent.stopPropagation();
        }
        if (segment.events.length === 0 && !segment.isDisabled && !segment.isCancelled) {
            this.segmentClicked.emit({ segment: segment });
        }
    }
};
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "title", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "day", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "segment", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "locale", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "customTemplate", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "hourSegmentHeight", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHourSegmentComponent.prototype, "showHour", void 0);
tslib_1.__decorate([
    Output()
], CalendarSchedulerHourSegmentComponent.prototype, "segmentClicked", void 0);
CalendarSchedulerHourSegmentComponent = tslib_1.__decorate([
    Component({
        selector: 'calendar-scheduler-hour-segment',
        template: `
        <ng-template #defaultTemplate>
            <div class="cal-scheduler-hour-segment"
                [title]="title"
                [ngClass]="segment?.cssClass"
                [class.has-events]="segment.events.length > 0"
                [class.cal-cancelled]="segment.isCancelled"
                [class.cal-disabled]="segment.isDisabled"
                [style.backgroundColor]="segment.backgroundColor"
                [style.height.px]="hourSegmentHeight"
                (mwlClick)="onSegmentClick($event, segment)">
                <div class="cal-scheduler-time unselectable" *ngIf="showHour && segment.events.length === 0">
                    {{ segment.date | calendarDate:'dayViewHour':locale }}
                </div>
            </div>
        </ng-template>
        <ng-template
            [ngTemplateOutlet]="customTemplate || defaultTemplate"
            [ngTemplateOutletContext]="{
                title: title,
                day: day,
                segment: segment,
                locale: locale,
                hourSegmentHeight: hourSegmentHeight,
                showHour: showHour,
                segmentClicked: segmentClicked
            }">
        </ng-template>
    `
    })
], CalendarSchedulerHourSegmentComponent);
export { CalendarSchedulerHourSegmentComponent };
