// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-event-content.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ElementRef, AfterViewInit } from '@angular/core';
let CalendarSchedulerEventContentComponent = class CalendarSchedulerEventContentComponent {
    constructor(hostElement) {
        this.hostElement = hostElement;
    }
    ngAfterViewInit() {
        setTimeout(() => { this.maxHeight = this.eventContainer.clientHeight - 70; }, 0);
    }
};
CalendarSchedulerEventContentComponent.ctorParameters = () => [
    { type: ElementRef }
];
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventContentComponent.prototype, "event", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventContentComponent.prototype, "eventContainer", void 0);
CalendarSchedulerEventContentComponent = tslib_1.__decorate([
    Component({
        selector: 'calendar-scheduler-event-content',
        template: `
        <div *ngIf="event.content"
            class="cal-scheduler-event-content"
            [style.max-height.px]="maxHeight"
            [innerHTML]="maxHeight && maxHeight >= 30 ? event.content : null">
        </div>
    `,
        host: {
            'class': 'cal-scheduler-event-content-container'
        }
    })
], CalendarSchedulerEventContentComponent);
export { CalendarSchedulerEventContentComponent };
