// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-event-title.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
let CalendarSchedulerEventTitleComponent = class CalendarSchedulerEventTitleComponent {
    constructor() {
        this.showStatus = true;
        this.showContent = true;
        this.showActions = true;
    }
};
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "view", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "event", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "showStatus", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "showContent", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "showActions", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventTitleComponent.prototype, "customTemplate", void 0);
CalendarSchedulerEventTitleComponent = tslib_1.__decorate([
    Component({
        selector: 'calendar-scheduler-event-title',
        template: `
        <ng-template #defaultTemplate>
            <div
                class="cal-scheduler-event-title"
                [innerHTML]="event.title | schedulerEventTitle:view:event">
            </div>
            <calendar-scheduler-event-actions
                *ngIf="showActions && !showContent && (event.isClickable || event.isDisabled)"
                class="no-content-actions"
                [event]="event">
            </calendar-scheduler-event-actions>
            <div *ngIf="event.status && showStatus"
                class="cal-scheduler-event-status"
                [class.ok]="event.status === 'ok'"
                [class.warning]="event.status === 'warning'"
                [class.danger]="event.status === 'danger'">
            </div>
        </ng-template>
        <ng-template
            [ngTemplateOutlet]="customTemplate || defaultTemplate"
            [ngTemplateOutletContext]="{
                view: view,
                event: event,
                showStatus: showStatus,
                showContent: showContent,
                showActions: showActions
            }">
        </ng-template>
    `,
        host: {
            'class': 'cal-scheduler-event-title-container'
        }
    })
], CalendarSchedulerEventTitleComponent);
export { CalendarSchedulerEventTitleComponent };
