// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-header.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
let CalendarSchedulerHeaderComponent = class CalendarSchedulerHeaderComponent {
    constructor() {
        this.dayHeaderClicked = new EventEmitter();
    }
    /**
     * @hidden
     */
    onDayHeaderClick(mouseEvent, day) {
        if (mouseEvent.stopPropagation) {
            mouseEvent.stopPropagation();
        }
        this.dayHeaderClicked.emit({ day: day });
    }
};
tslib_1.__decorate([
    Input()
], CalendarSchedulerHeaderComponent.prototype, "days", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHeaderComponent.prototype, "locale", void 0);
tslib_1.__decorate([
    Input()
], CalendarSchedulerHeaderComponent.prototype, "customTemplate", void 0);
tslib_1.__decorate([
    Output()
], CalendarSchedulerHeaderComponent.prototype, "dayHeaderClicked", void 0);
CalendarSchedulerHeaderComponent = tslib_1.__decorate([
    Component({
        selector: 'calendar-scheduler-header',
        template: `
        <ng-template #defaultTemplate>
            <div class="cal-scheduler-headers">
                <div class="cal-header aside cal-header-clock align-center">
                    <i class="material-icons md-32" style="margin:auto;">schedule</i>
                </div>

                <div class="cal-header-cols aside">
                    <div
                        class="cal-header"
                        *ngFor="let day of days"
                        [class.cal-past]="day.isPast"
                        [class.cal-today]="day.isToday"
                        [class.cal-future]="day.isFuture"
                        [class.cal-weekend]="day.isWeekend"
                        (mwlClick)="onDayHeaderClick($event, day)">
                        <b>{{ day.date | calendarDate:'weekViewColumnHeader':locale }}</b><br>
                        <span>{{ day.date | calendarDate:'weekViewColumnSubHeader':locale }}</span>
                    </div>
                </div>
            </div>
        </ng-template>
        <ng-template
            [ngTemplateOutlet]="customTemplate || defaultTemplate"
            [ngTemplateOutletContext]="{
                days: days,
                locale: locale,
                dayHeaderClicked: dayHeaderClicked
            }">
        </ng-template>
    `
    })
], CalendarSchedulerHeaderComponent);
export { CalendarSchedulerHeaderComponent };
