// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-event-actions.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
let CalendarSchedulerEventActionsComponent = class CalendarSchedulerEventActionsComponent {
    constructor() {
        this.actions = [];
    }
    ngOnInit() {
        this.setupActions();
    }
    ngOnChanges(changes) {
        if (changes.event) {
            this.setupActions();
        }
    }
    setupActions() {
        this.actions = this.event.isCancelled
            ? this.event.actions.filter((a) => !a.when || a.when === 'cancelled')
            : this.event.isDisabled
                ? this.event.actions.filter((a) => !a.when || a.when === 'disabled')
                : this.event.actions.filter((a) => !a.when || a.when === 'enabled');
    }
    /**
     * @hidden
     */
    onActionClick(mouseEvent, action, event) {
        if (mouseEvent.stopPropagation) {
            mouseEvent.stopPropagation();
        }
        action.onClick(event);
    }
};
tslib_1.__decorate([
    Input()
], CalendarSchedulerEventActionsComponent.prototype, "event", void 0);
CalendarSchedulerEventActionsComponent = tslib_1.__decorate([
    Component({
        selector: 'calendar-scheduler-event-actions',
        template: `
        <span *ngIf="event.actions"
            class="cal-scheduler-event-actions">
            <a
                class="cal-scheduler-event-action"
                href="javascript:;"
                *ngFor="let action of actions"
                (mwlClick)="onActionClick($event, action, event)"
                [ngClass]="action.cssClass"
                [innerHtml]="action.label"
                [title]="action.title">
            </a>
        </span>
    `,
        host: {
            'class': 'cal-scheduler-event-actions-container'
        }
    })
], CalendarSchedulerEventActionsComponent);
export { CalendarSchedulerEventActionsComponent };
