// angular-calendar-scheduler/esm5/modules/scheduler/pipes/calendar-scheduler-date.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, LOCALE_ID, Inject } from '@angular/core';
import { CalendarDateFormatter, DateAdapter } from 'angular-calendar';
var CalendarSchedulerDatePipe = /** @class */ (function () {
    function CalendarSchedulerDatePipe(dateAdapter, dateFormatter, locale) {
        this.dateAdapter = dateAdapter;
        this.dateFormatter = dateFormatter;
        this.locale = locale;
    }
    CalendarSchedulerDatePipe.prototype.transform = function (date, method, locale, weekStartsOn, excludeDays, daysInWeek, startsWithToday) {
        if (locale === void 0) { locale = this.locale; }
        if (weekStartsOn === void 0) { weekStartsOn = 0; }
        if (excludeDays === void 0) { excludeDays = []; }
        if (typeof this.dateFormatter[method] === 'undefined') {
            var allowedMethods = Object.getOwnPropertyNames(Object.getPrototypeOf(CalendarDateFormatter.prototype)).filter(function (iMethod) { return iMethod !== 'constructor'; });
            throw new Error(method + " is not a valid date formatter. Can only be one of " + allowedMethods.join(', '));
        }
        return this.dateFormatter[method]({
            dateAdapter: this.dateAdapter,
            date: date,
            locale: locale,
            weekStartsOn: weekStartsOn,
            excludeDays: excludeDays,
            daysInWeek: daysInWeek,
            startsWithToday: startsWithToday
        });
    };
    CalendarSchedulerDatePipe.ctorParameters = function () { return [
        { type: DateAdapter },
        { type: CalendarDateFormatter },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    CalendarSchedulerDatePipe = tslib_1.__decorate([
        Pipe({
            name: 'calendarSchedulerDate'
        }),
        tslib_1.__param(2, Inject(LOCALE_ID))
    ], CalendarSchedulerDatePipe);
    return CalendarSchedulerDatePipe;
}());
export { CalendarSchedulerDatePipe };
